File: ruby/grade-school/grade_school.rb
class School

  def initialize
    @students = []
  end

  def to_hash
    hash_students = @students.sort!.to_h
    hash_students.hash_reverse.sort_by {|k| k}.to_h
  end

  def add(name, grade)
    ary = []
    ary << grade
    ary << name
    @students << ary.reverse
  end

  def grade(k)
    @students.inject([]) do |ary, student|
      ary << student.first if student.last == 5
      ary.sort
    end
  end
end

  class Hash
    def hash_reverse
      each_with_object({}) do |(k, v), ary|
        ary[v] ||= []
        ary[v] << k
      end
    end
  end

File: javascript/grade-school/grade-school.js
module.exports = function() {
  var roster = {};

  function attrRoster() {
    return roster;
  }

  function addStudents(name, grade) {
    roster[grade] ? roster[grade].push(name) : roster[grade] = [name];
    sortRoster();
  }

  function studentsByGrade(grade) {
    return roster[grade] ? roster[grade].sort() : [];
  }

  function sortRoster() {
    for(grade in roster) {
      return roster[grade] = roster[grade].sort();
    }
  }

  return {
    add: addStudents,
    roster: attrRoster,
    grade: studentsByGrade
  }
}
